mycroft/skills/intent_services/adapt_service.py

```diff
@@ -0,0 +1,229 @@
+                    result.append(entity)
+                    # NOTE: this implies that we will only ever get one
+                    # of an entity kind from context, unless specified
+                    # multiple times in missing_entities. Cannot get
+                    # an arbitrary number of an entity kind.
+                    missing_entities.remove(entity.get('data'))
+        else:
+            result = context
+
+        # Only use the latest instance of each keyword
+        stripped = []
+        processed = []
+        for f in result:
+            keyword = f['data'][0][1]
+            if keyword not in processed:
+                stripped.append(f)
+                processed.append(keyword)
+        result = stripped
+        return result
+
+
+class AdaptService:
+    def __init__(self, config):
+        self.config = config
+        self.engine = IntentDeterminationEngine()
+        # Context related intializations
+        self.context_keywords = self.config.get('keywords', [])
+        self.context_max_frames = self.config.get('max_frames', 3)
+        self.context_timeout = self.config.get('timeout', 2)
+        self.context_greedy = self.config.get('greedy', False)
+        self.context_manager = ContextManager(self.context_timeout)
+
+    def update_context(self, intent):
+        """Updates context with keyword from the intent.
+
+        NOTE: This method currently won't handle one_of intent keywords
+              since it's not using quite the same format as other intent
+              keywords. This is under investigation in adapt, PR pending.
+
+        Args:
+            intent: Intent to scan for keywords
+        """
+        for tag in intent['__tags__']:
+            if 'entities' not in tag:
+                continue
+            context_entity = tag['entities'][0]
+            if self.context_greedy:
+                self.context_manager.inject_context(context_entity)
+            elif context_entity['data'][0][1] in self.context_keywords:
+                self.context_manager.inject_context(context_entity)
+
+    def match_intent(self, utterances, lang, _=None):
+        """Run the Adapt engine to search for an matching intent.
+
+        Arguments:
+            utterances (iterable): iterable of utterances, expected order
+                                   [raw, normalized, other]
+            lang (string): language code, e.g "en-us"
+
+        Returns:
+            Intent structure, or None if no match was found.
+        """
+        best_intent = None
+
+        def take_best(intent, utt):
+            nonlocal best_intent
+            best = best_intent.get('confidence', 0.0) if best_intent else 0.0
+            conf = intent.get('confidence', 0.0)
+            if conf > best:
+                best_intent = intent
+                # TODO - Shouldn't Adapt do this?
+                best_intent['utterance'] = utt
+
+        for idx, utt in enumerate(utterances):
```

Comment:
Does this loop need to be enumerated? I can't see the index being used, is there another benefit to enumeration here?

Reply:
I may need to do a rethink, the idx was used to map normalized to raw utterances making sure the raw utterance was reported to the skill...It should still do that...

I might reorganize slightly to send a list of tuples instead of a flat list.
```
[(raw_utt, (optional)normalized_utt)]
```

Reply:
I pushed the WIP rethink...makes sure the original is always sent along from adapt like previously and removes the enumeration here...

I'm thinking a bit if Padatious should do this as well...

Reply:
Forgive my naivety, what's the thinking behind the list of tuples? Is the idea that the intent service might at some point receive multiple distinct utterances and need to select the best intent from both? Or there's no real use case at the moment and it's more of a just in case thing?

Reply:
The list was added way back in the long ago when we wanted to support multiple hypotheses from the STT. so this is basically to support that if it ever gets implemented.

Reply:
In the _original_ original implementation, the recognizer_loop would perform wakeword detection, and then extract the wakeword phrase from the recognized text. The intent engine would take the (potentially 2) parts, and evaluate them independently (dropping useless parts on the ground, as you do). This allowed for interactions like

```
I can't stand the quiet. Mycroft, turn on some music
```
 as well as a trailing wakeword:
```
Turn on some music, mycroft
```

Unfortunately, it was _blasted_ hard to get right, and we moved back to a prefix-wakeword model.